src/monodroid/jni/jni-wrappers.h

```diff
@@ -0,0 +1,174 @@
+// Dear Emacs, this is a -*- C++ -*- header
+#ifndef __JNI_WRAPPERS_H
+#define __JNI_WRAPPERS_H
+
+#include <assert.h>
+#include <jni.h>
+
+#ifdef __cplusplus
+#include <utility>
+
+namespace xamarin { namespace android
+{
+	class jstring_array_wrapper;
+
+	class jstring_wrapper
+	{
+	public:
+		explicit jstring_wrapper (JNIEnv *env) noexcept
+			: env (env),
+			  jstr (nullptr)
+		{
+			assert (env);
+		}
+
+		explicit jstring_wrapper (JNIEnv *env, const jobject jo) noexcept
+			: env (env),
+			  jstr (reinterpret_cast<jstring> (jo))
+		{
+			assert (env);
+		}
+
+		explicit jstring_wrapper (JNIEnv *env, const jstring js) noexcept
+			: env (env),
+			  jstr (js)
+		{
+			assert (env);
+		}
+
+		~jstring_wrapper () noexcept
+		{
+			release ();
+		}
+
+		const char* get_cstr () noexcept
+		{
+			if (cstr == nullptr && env != nullptr)
+				cstr = env->GetStringUTFChars (jstr, nullptr);
+
+			return cstr;
+		}
+
+		jstring_wrapper& operator= (const jobject new_jo) noexcept
+		{
+			assign (reinterpret_cast<jstring> (new_jo));
+			return *this;
+		}
+
+		jstring_wrapper& operator= (jstring new_js) noexcept
+		{
+			assign (new_js);
+			return *this;
+		}
+
+	protected:
+		jstring_wrapper ()
+			: env (nullptr),
+			  jstr (nullptr)
+		{}
+
+		void release () noexcept
+		{
+			if (jstr == nullptr || cstr == nullptr || env == nullptr)
+				return;
+			env->ReleaseStringUTFChars (jstr, cstr);
+			jobjectRefType type = env->GetObjectRefType (jstr);
+			switch (type) {
+				case JNILocalRefType:
+					env->DeleteLocalRef (jstr);
+					break;
+
+				case JNIGlobalRefType:
+					env->DeleteGlobalRef (jstr);
+					break;
+
+				case JNIWeakGlobalRefType:
+					env->DeleteWeakGlobalRef (jstr);
+					break;
+
+				case JNIInvalidRefType: // To hush compiler warning
+					break;
+			}
+
+			jstr = nullptr;
+			cstr = nullptr;
+		}
+
+		void assign (jstring new_js) noexcept
+		{
+			release ();
+			if (new_js == nullptr)
+				return;
+
+			jstr = new_js;
+			cstr = nullptr;
+		}
+
+		friend class jstring_array_wrapper;
+
+	private:
+		JNIEnv *env;
+		jstring jstr;
+		const char *cstr = nullptr;
+	};
+
+	class jstring_array_wrapper
+	{
+	public:
+		explicit jstring_array_wrapper (JNIEnv *env) noexcept
+			: env (env),
+			  arr (nullptr),
+			  len (0)
+		{
+			assert (env);
+		}
+
+		explicit jstring_array_wrapper (JNIEnv *env, jobjectArray arr)
+			: env (env),
+			  arr (arr)
+		{
+			assert (env);
+			assert (arr);
+			len = env->GetArrayLength (arr);
+			if (len > sizeof (static_wrappers) / sizeof (jstring_wrapper))
+				wrappers = new jstring_wrapper [len];
+			else
+				wrappers = static_wrappers;
+		}
+
+		~jstring_array_wrapper () noexcept
+		{
+			if (wrappers != nullptr && wrappers != static_wrappers)
+				delete[] wrappers;
+		}
+
+		size_t get_length () const noexcept
```

Comment:
Shouldn't this be called `size()`, for consistency with e.g. `std:vector` and other collections?

Reply:
length != size. **Length** of the array is 10 elements, but its **size** will be times the pointer size. I don't want to mimic STL, especially where it's wrong.

Reply:
ROFLMAO.